modules/modal/react/lib/Modal.tsx

```diff
@@ -69,125 +69,124 @@ const transformOrigin = {
+      // Setting tabIndex allows the header to be focusable.
+      // We do the header instead of the next focusable element to prevent useful context from being skipped
+      firstFocusable.tabIndex = 0;
+      firstFocusable.style.outline = 'none';
+
+      const changeTabIndex = () => {
+        // We no longer need to focus on the header after it loses focus
+        // We simply want to transfer focus inside
+        firstFocusable.removeEventListener('blur', changeTabIndex);
+        // We must wait one frame to ensure tabbable checks are satisfied
+        // by all focus libraries...
+        requestAnimationFrame(() => {
+          firstFocusable.removeAttribute('tabIndex');
+        });
+      };
+      firstFocusable.addEventListener('blur', changeTabIndex);
     }
+    return firstFocusable;
+  } else {
+    throw new Error(
+      'No focusable element was found. Please ensure modal has at least one focusable element'
+    );
   }
 }
 
-export default class Modal extends React.Component<ModalProps> {
-  private modalRef = React.createRef<HTMLDivElement>();
-
-  static Padding = PopupPadding;
-  static Width = ModalWidth;
-  static defaultProps = {
-    open: false,
-    padding: Modal.Padding.l,
-    width: Modal.Width.s,
-    closeOnEscape: true,
-  };
+const useKeyDownListener = (handleKeydown: EventListenerOrEventListenerObject) => {
+  React.useEffect(() => {
+    document.addEventListener('keydown', handleKeydown);
+    return () => {
+      document.removeEventListener('keydown', handleKeydown);
+    };
+  }, []);
+};
+
+const useInitialFocus = (
+  modalRef: React.RefObject<HTMLElement>,
+  firstFocusableRef: React.RefObject<HTMLElement> | undefined
+) => {
+  React.useEffect(() => {
+    if (modalRef.current) {
+      const elem =
+        (firstFocusableRef && firstFocusableRef.current) ||
+        getFirstElementToFocus(modalRef.current);
+      elem.focus();
+    }
+  }, [modalRef]);
+};
+
+const Modal = ({
+  open,
+  handleClose,
+  padding,
+  width,
+  heading,
+  children,
+  closeOnEscape,
+  firstFocusRef,
+  ...elemProps
+}: ModalProps): JSX.Element | null => {
+  const modalRef = React.useRef<HTMLDivElement>(null);
+
+  const handleKeydown = (event: KeyboardEvent) => {
+    if (modalRef.current) {
+      tabTrappingKey(event, modalRef.current);
+    }
 
-  private handleKeydown = (event: KeyboardEvent) => {
-    if (
-      this.props.closeOnEscape &&
-      this.props.handleClose &&
-      (event.key === 'Esc' || event.key === 'Escape')
-    ) {
-      this.props.handleClose();
+    if (closeOnEscape && handleClose && (event.key === 'Esc' || event.key === 'Escape')) {
+      handleClose();
     }
   };
 
-  componentDidMount() {
-    document.addEventListener('keydown', this.handleKeydown);
-  }
-
-  componentWillUnmount() {
-    document.removeEventListener('keydown', this.handleKeydown);
-  }
+  useKeyDownListener(handleKeydown);
+  useInitialFocus(modalRef, firstFocusRef);
```

Comment:
We haven't released a component that depends on React hooks yet. Not all our dependents are using React 16.8+ yet. Hooks in this file have been thus far optional. Are we ready to break compatibility @anicholls?

Reply:
hmm okay, I just saw that canvas is using 16.8.6. What dependencies are blocking us to use hooks?